File: tests/perl5/test-RNA-constraints-SHAPE.pl
#!/usr/bin/perl -I.libs

######################### We start with some black magic to print on failure.
# (It may become useful if the test is moved to ./t subdirectory.)
use strict;
use Test::More tests => 4;
use Data::Dumper;
use FileHandle;

use RNA;
use warnings;
use RNAHelpers qw(:Paths);


my $datadir = getDataDirPath();

my $seq_con     = "CCCAAAAGGGCCCAAAAGGG";
my $str_con     = "..........(((....)))";
my $str_con_def = "(((....)))(((....)))";
my $fc; #fold compound reference
my $ss=""; #return string
my $mfe=0;
my $energy=0;



sub getShapeDataFromFile
{
    my $filepath = shift(@_);
    my @retVec;
    push(@retVec,-999.0);# data list is 1-based, so we add smth. at pos 0
    my $count=1;
    open(my $fh, "<", $filepath) || die "Couldn't open '".$filepath."' for reading because: ".$!;

    foreach my $line (<$fh>)
    {

        my @sp= split(/ /,$line);
        my $pos = $sp[0];
        my $value = $sp[1];

        if($pos == $count)
        {
            push(@retVec,$value +0.00); # make float from string
        }else
        {
            for(my $i=$pos;$i<=$count;$i++)
            {
                push(@retVec,-999.0);
            }
            push(@retVec,$value);
            $count=$pos;
        }
        $count+=1;

    }
    return @retVec;
}

sub getShapeSequenceFromFile
{
    my $f = shift(@_);
    my $retSeq ="";
    open(my $fh, "<", $f) || die "Couldn't open '".$f."' for reading because: ".$!;
    foreach my $line (<$fh>)
    {
        chomp $line;
        return $line; #return the first line
    }
}

##################################
##test_sc_add_deigan
print "test_sc_add_deigan\n";
my $seq  =  getShapeSequenceFromFile($datadir . "Lysine_riboswitch_T._martima.db");
my @reactivities = getShapeDataFromFile($datadir . "Lysine_riboswitch_T._martima.shape_2rows");

$fc= new RNA::fold_compound($seq);
print "@reactivities \n";

$fc->sc_add_SHAPE_deigan(\@reactivities,1.9,-0.7,RNA::OPTION_DEFAULT);
($ss,$mfe) = $fc->mfe();
printf("%s [%6.2f] \n",$ss,$mfe);
is(sprintf ("%6.2f",$mfe), sprintf("%6.2f",-121.55));
##################################
##test_sc_add_SHAPE_deigan2
print "test_sc_add_SHAPE_deigan2";
my $seq_ribo  =  getShapeSequenceFromFile($datadir . "TPP_riboswitch_E.coli.db");
my @reactivities_ribo = getShapeDataFromFile($datadir . "TPP_riboswitch_E.coli.shape_2rows");

$fc= new RNA::fold_compound($seq_ribo);
print "@reactivities_ribo \n";

$fc->sc_add_SHAPE_deigan(\@reactivities_ribo,1.9,-0.7,RNA::OPTION_DEFAULT);
($ss,$mfe) = $fc->mfe();
printf("%s [%6.2f] \n",$ss,$mfe);
is(sprintf ("%6.2f",$mfe), sprintf("%6.2f",-52.61));
##################################
# I just added completely randomly choosen sequences and shape data, this test only checks if the function can be called, not if it results correct results.
##test_sc_add_SHAPE_deigan_ali
print "test_sc_add_SHAPE_deigan_ali";
# shape data from
my $shapeSeq1 = "CCCAAAAGGG";
my $shapeSeq2 = "AAUAAAAAUU";

#my @shapeData1 = (-999.0,0.04,1.12,1.1,-999.0,0.05,0.5,0.3,-999.0,1.4);
#my @shapeData2 = (-999.0,-999.0,-999.0,1.23,1.4,0.05,0.5,0.3,-999.0,1.4);
my @shapeAli = ($shapeSeq1,$shapeSeq2);
$fc= new RNA::fold_compound(\@shapeAli);

my @assoc = (-1,1,2);
my $ret = $fc->sc_add_SHAPE_deigan_ali(\@shapeAli, \@assoc,1.8,-0.6);
($ss,$mfe) = $fc->mfe();
printf("%s [%6.2f] \n",$ss,$mfe);
is($ret,1);
##################################
##test_sc_add_SHAPE_zarringhalam
print("test_sc_add_SHAPE_zarringhalam");
$seq_ribo  =  getShapeSequenceFromFile($datadir . "TPP_riboswitch_E.coli.db");
@reactivities_ribo = getShapeDataFromFile($datadir . "TPP_riboswitch_E.coli.shape_2rows");

$fc= new RNA::fold_compound($seq_ribo);
print "@reactivities_ribo \n";

$fc->sc_add_SHAPE_zarringhalam(\@reactivities_ribo,0.5,0.5,"O");
($ss,$mfe) = $fc->mfe();
printf("%s [%6.2f] \n",$ss,$mfe);
is(sprintf ("%6.2f",$mfe), sprintf("%6.2f",-5.28));

undef $fc;

File: tests/python/test_RNA-constraints-SHAPE.py
import os
import unittest

if __name__ == '__main__':
    from py_include import taprunner
    import RNApath
    RNApath.addSwigInterfacePath()

import RNA


seq_con     = "CCCAAAAGGGCCCAAAAGGG"
str_con     = "..........(((....)))"
str_con_def = "(((....)))(((....)))"

def getShapeDataFromFile(filepath):
    retVec = []
    retVec.append(-999.0);  # data list is 1-based, so we add smth. at pos 0
    count=1
    with open(filepath, 'r') as f:
        lines = f.read().splitlines()

        for line in lines:
            pos = int(line.split(' ')[0])
            value = float(line.split(' ')[1])

            if(pos==count):
                retVec.append(value)
            else:
                for i in range(pos-count):
                    retVec.append(-999.0)
                retVec.append(value)
                count=pos
            count+=1
    return retVec


def getShapeSequenceFromFile(filepath):
    retSeq=""
    with open(filepath,'r') as f:
        lines = f.read().splitlines()

    return lines[0]

class constraintsTest(unittest.TestCase):
    DATADIR = os.environ.get('VRNA_TEST_DATA', os.sep.join(["tests", "data"]))

    def test_sc_add_deigan(self):
        """SHAPE data Deigan et al. 2009 method - Lysine riboswitch"""
        seq          = getShapeSequenceFromFile(os.sep.join([self.DATADIR, "Lysine_riboswitch_T._martima.db"]))
        reactivities = getShapeDataFromFile(os.sep.join([self.DATADIR, "Lysine_riboswitch_T._martima.shape_2rows"]))

        fc=RNA.fold_compound(seq)
        print(reactivities)

        ret = fc.sc_add_SHAPE_deigan(reactivities, 1.9, -0.7, RNA.OPTION_DEFAULT)
        (ss,mfe) = fc.mfe()
        print(ss, "[ %6.4f ]" % mfe)
        self.assertEqual("%6.4f" %mfe,"%6.4f" % -121.55 )


    def test_sc_add_SHAPE_deigan2(self):
        """SHAPE data Deigan et al. 2009 method - TPP riboswitch"""
        seq_ribo     = getShapeSequenceFromFile(os.sep.join([self.DATADIR, "TPP_riboswitch_E.coli.db"]))
        reactivities = getShapeDataFromFile(os.sep.join([self.DATADIR, "TPP_riboswitch_E.coli.shape_2rows"]))

        fc=RNA.fold_compound(seq_ribo)
        print(reactivities)

        ret = fc.sc_add_SHAPE_deigan(reactivities,1.9,-0.7,RNA.OPTION_DEFAULT);  # these values were copied from ronnys Talk about constraints
        (ss,mfe) = fc.mfe()
        print(ss, "[ %6.2f ]" % mfe)
        self.assertEqual("%6.2f" %mfe,"%6.2f" % -52.61 )


    # with randomly choosen sequences and shape data, this test only checks if the function can be called, not if it results correct results.
    def test_sc_add_SHAPE_deigan_ali(self):
        """SHAPE data Deigan et al. 2009 method - Comparative MFE prediction"""
        # shape data from
        shapeSeq1 = "CCCAAAAGGG"
        shapeSeq2 = "AAUAAAAAUU"
        shapeAli = [shapeSeq1,shapeSeq2]

        shapeFiles = [os.sep.join([self.DATADIR, "alignment_1.shape_2rows"]), os.sep.join([self.DATADIR, "alignment_2.shape_2rows"])]
        fc=RNA.fold_compound(shapeAli);
        assoc = [0, 1]
        ret = fc.sc_add_SHAPE_deigan_ali(shapeFiles, assoc,1.8,-0.6)
        (ss,mfe) = fc.mfe()
        print(ss, "[ %6.2f ]" % mfe)
        self.assertEqual(ret,2)


    def test_sc_add_SHAPE_zarringhalam(self):
        """SHAPE data Zarringhalam et al 2012 method"""
        seq_ribo     = getShapeSequenceFromFile(os.sep.join([self.DATADIR, "TPP_riboswitch_E.coli.db"]))
        fc           = RNA.fold_compound(seq_ribo)
        reactivities = getShapeDataFromFile(os.sep.join([self.DATADIR, "TPP_riboswitch_E.coli.shape_2rows"]))

        ret = fc.sc_add_SHAPE_zarringhalam(reactivities, 0.5, 0.5, "O"); # these values were copied from ronnys Talk about constraints, O = default value
        (ss,mfe) = fc.mfe()
        print(ss, "[ %6.2f ]" % mfe)
        self.assertEqual("%6.2f" %mfe,"%6.2f" % -5.28 )


if __name__ == '__main__':
    constraintsTest.DATADIR = RNApath.getDataDirPath()
    unittest.main(testRunner=taprunner.TAPTestRunner())
